bin/gql-gen.ts

```diff
@@ -40,53 +40,54 @@ async function main(): Promise<void> {
         const cwd = process.cwd();
         const lib = libDir(cwd);
 
+        // patch up codegen template to create optional properties
+        const codegenTemplate = path.join(cwd, "node_modules", "graphql-codegen-typescript-client", "dist", "index.js");
+        if (await fs.pathExists(codegenTemplate)) {
+            let contents = (await fs.readFile(codegenTemplate)).toString();
+            contents = contents.replace(
+                /{{ name }}: {{ convertedFieldType/i,
+                "{{ name }}?: {{ convertedFieldType");
+            await fs.writeFile(codegenTemplate, contents);
+        }
+
         // check if the project has a custom schema
         const customSchemaLocation = path.join(lib, "graphql", "schema.json");
         const defaultSchemaLocation = path.join(cwd, "node_modules", "@atomist", "automation-client", "lib",
             "graph", "schema.json");
         const schema = fs.existsSync(customSchemaLocation) ? customSchemaLocation : defaultSchemaLocation;
 
-        const gqlGenCmd = path.join(cwd, "node_modules", ".bin", "gql-gen") +
-            ((process.platform === "win32") ? ".cmd" : "");
         const gqlGenOutput = path.join(lib, "typings", "types.ts");
-        const gqlGenArgs = [
-            "--schema", schema,
-            "--skip-schema",
-            "--template", "graphql-codegen-typescript-template",
-            "--out", gqlGenOutput,
-        ];
-
-        const opts: child_process.SpawnOptions = {
-            cwd,
-            env: process.env,
-            stdio: "inherit",
-        };
+        await fs.ensureDir(path.dirname(gqlGenOutput));
 
         const graphQlGlob = `${lib}/graphql/!(ingester)/*.graphql`;
 
+        const config: Types.Config = {
+            overwrite: true,
+            schema: [schema],
+            generates: {
+                [gqlGenOutput]: {
+                    plugins: [
+                        "typescript-common",
+                        "typescript-client",
+                    ],
+                    config: {
+                        namingConvention: {
+                            enumValues: "keep",
+                        },
+                    },
+                },
+            },
+        };
+
         const graphqlFiles = await util.promisify(glob)(graphQlGlob);
+
         if (graphqlFiles && graphqlFiles.length > 0) {
-            gqlGenArgs.push(graphQlGlob);
+            config.documents = [graphQlGlob];
+            await generate(config);
         } else {
-            console.info("No GraphQL files found in project, generating default types");
+            console.info("No GraphQL files found in project. Skipping type generation...");
```

Comment:
Are the default types brought in via a dependency?

Reply:
What default types? Without any graphql files, what types would you expect?

Reply:
All the core types in the graph. Typically when there are no custom types in the project, you still get a types.ts with all core types Atomist defines. Will those be available some other way.

Reply:
No. They won’t.

Reply:
You only get a generated types.ts when you actually define graphql queries or subscriptions. At that point you get a types.ts with the core types you used in your queries/subscriptions.

Reply:
I think I am confusing custom schema and GraphQL files. You still create a types.ts if there is no custom schema, but you do not if there are no GraphQL queries/subscriptions/etc. Sorry for the confusion.